### ISSUE TITLE
Contribute our local fix upstream to Python

### ISSUE THREAD

**Issue by CleanCut** [CleanCut/green#215]
[These lines](https://github.com/CleanCut/green/commit/c988919561ea1a83e3d4d5ff4a99663b554fc2d3#diff-3bb96ef79f5c5b3bb6bcf1ae5e1815e3R105-R109
) are a monkey-patch I made for unittest to work around an upstream bug that was reported in https://github.com/CleanCut/green/issues/159

Let's contribute that fix upstream to Python!  Then maybe we could _not_ monkey patch our stuff.

I'm pretty busy, so if someone else wants to do the honors, please go ahead.  If not, I'll add myself as an "assignee" when I finally get around to it.

**Comment by CleanCut** [comment]
I finally sat down to do this today and discovered that this was fixed as part of https://github.com/python/cpython/pull/9190 that was included in Python 3.8.

The new task for this issue is:
- [ ] Only do the monkey-patch for Python versions less than 3.8

**Comment by CleanCut** [comment]
Hmm...it appears the different way that they fixed the problem interacts badly with how we count up skipped tests...so I'll have to look into that. 🤦‍♂ 

- [ ] Adapt to upstream's SkipTest fix in Python 3.8+

```python
Failure in green.test.test_suite.TestFunctional.test_skip_in_setUpClass
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/travis/build/CleanCut/green/green/test/test_suite.py", line 157, in test_skip_in_setUpClass
    self.assertEqual(len(result.skipped), 2)
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 0 != 2

###

Error in green.test.test_suite.TestGreenTestSuite.test_skip_in_setUpClass
  File "/opt/python/3.8.0/lib/python3.8/unittest/suite.py", line 163, in _handleClassSetUp
    setUpClass()
  File "/opt/python/3.8.0/lib/python3.8/unittest/mock.py", line 1075, in __call__
    return self._mock_call(*args, **kwargs)
  File "/opt/python/3.8.0/lib/python3.8/unittest/mock.py", line 1079, in _mock_call
    return self._execute_mock_call(*args, **kwargs)
  File "/opt/python/3.8.0/lib/python3.8/unittest/mock.py", line 1134, in _execute_mock_call
    raise effect
unittest.case.SkipTest: kaboom

During handling of the above exception, another exception occurred:
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/python/3.8.0/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/travis/build/CleanCut/green/green/test/test_suite.py", line 104, in test_skip_in_setUpClass
    gts._handleClassSetUp(mock_test, mock_result)
  File "/opt/python/3.8.0/lib/python3.8/unittest/suite.py", line 168, in _handleClassSetUp
    className = util.strclass(currentClass)
  File "/opt/python/3.8.0/lib/python3.8/unittest/util.py", line 55, in strclass
    return "%s.%s" % (cls.__module__, cls.__qualname__)
  File "/opt/python/3.8.0/lib/python3.8/unittest/mock.py", line 637, in __getattr__
    raise AttributeError(name)
AttributeError: __qualname__
```

**Comment by CleanCut** [comment]
Found and fixed the problem in `green.test.test_suite.TestGreenTestSuite.test_skip_in_setUpClass` -- the test itself broke due to changes in internal unittest implementation in Python >= 3.8.

I suspect the other failure is a similar failure due to reliance on unittest internals. Hopefully I can track that down as well.

**Comment by CleanCut** [comment]
> I finally sat down to do this today and discovered that this was fixed as part of [python/cpython#9190](https://github.com/python/cpython/pull/9190) that was included in Python 3.8.

Nope, I was completely wrong.  The change in Python 3.8 did _not_ fix the problem, it just changed code that broke our monkey patch to the problem.  I added a new, updated monkey-patch in 9f393edcaacc838bf3635f84a83537a6c7ce045b

Now the task in this issue reverts to:
- [ ] Contribute our local fix upstream to Python
